### PR TITLE
Spiral Ports

### PR DESCRIPTION
**Pull request by maxcapodi78**
user can now create a spiral inscribed inside a face and, from that one, create a complex 3d port that cover the full surface of the object. Useful when distributed port has to be created

### CODE REVIEW

File: pyaedt/hfss.py [hunk 1]
```diff
@@ -1577,13 +1594,183 @@ def create_lumped_port_between_objects(
...
             else:
                 faces = self.modeler.primitives.get_object_faces(sheet_name)
-                self._create_port_terminal(faces[0], endobject, portname, iswaveport=False)
-            return portname
+                return self._create_port_terminal(faces[0], endobject, portname, iswaveport=False)
         return False
 
+    @aedt_exception_handler
+    def create_spiral_lumped_port(self, start_object, end_object, port_width=None):
+        """Created a spiral lumped port between two adjacent objects.
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1605]:
```suggestion
        """Create a spiral lumped port between two adjacent objects.
```

File: pyaedt/hfss.py [hunk 2]
```diff
@@ -1577,13 +1594,183 @@ def create_lumped_port_between_objects(
...
+        The two objects needs to be two adjacent identical faces and faces have to be polygon (not circle).
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1606]:
```suggestion
        The two objects must be two adjacent identical faces and faces have to be polygon (not circle).
```

File: pyaedt/hfss.py [hunk 3]
```diff
@@ -1577,13 +1594,183 @@ def create_lumped_port_between_objects(
...
 
+    @aedt_exception_handler
+    def create_spiral_lumped_port(self, start_object, end_object, port_width=None):
+        """Create a spiral lumped port between two adjacent objects.
+        The two objects must be two adjacent identical faces and faces have to be polygon (not circle).
+
+        Parameters
+        ----------
+        start_object : str or int or :class:`pyaedt.modeler.Object3d.Object3d`
+        end_object
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1611]:
```suggestion
        end_object: str or int or :class:`pyaedt.modeler.Object3d.Object3d`
```

**Review by Alberto-DM — APPROVED** [review]
It looks good to me! Approved.

**Review by MaxJPRey — APPROVED** [review]

File: pyaedt/hfss.py [hunk 4]
```diff
@@ -1577,13 +1593,214 @@ def create_lumped_port_between_objects(
...
+
+        Returns
+        -------
+        :class:`pyaedt.modules.Boundary.BoundaryObject`
+            Boundary object.
+
+        """
+        # fmt: off
+        assert "Terminal" in self.solution_type, "This method can be used only in Terminal Solutions."
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1620]:
```suggestion
        if not "Terminal" in self.solution_type:
            raise Exception("This method can be used only in Terminal Solutions.")
```

File: pyaedt/hfss.py [hunk 5]
```diff
@@ -1577,13 +1593,214 @@ def create_lumped_port_between_objects(
...
+            else:
+                vertex_coordinates = []
+                for v in closest_faces[0].vertices:
+                    vertex_coordinates.append(v.position)
+                segments_lengths = []
+                for vc in vertex_coordinates:
+                    segments_lengths.append(GeometryOperators.points_distance(vc, closest_faces[0].center))
+                spiral_width = min(segments_lengths) / 15
+                filling = 1.5
+                print(filling)
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1672]: @maxcapodi78 Can be removed. It must have been for debugging.
```suggestion
```

File: pyaedt/hfss.py [hunk 6]
```diff
@@ -1577,13 +1593,214 @@ def create_lumped_port_between_objects(
...
+                for v in closest_faces[0].vertices:
+                    face_bb[i] = min(face_bb[i], v.position[i])
+                    face_bb[i+3] = max(face_bb[i+3], v.position[i])
+            # get the ratio in 2D
+            bb_dim = [abs(face_bb[i]-face_bb[i+3]) for i in range(3) if abs(face_bb[i]-face_bb[i+3]) > 1e-12]
+            bb_ratio = max(bb_dim)/min(bb_dim)
+            if bb_ratio > 2:
+                spiral_width = min(bb_dim) / 12
+                filling = -0.2828*bb_ratio**2 + 3.4141*bb_ratio - 4.197
+                print(filling)
```
Comment (MaxJPRey) [inline pyaedt/hfss.py:1662]:
```suggestion
```

**Pull request merged**